cmd/atlasci/internal/ci/ci.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2021-present The Atlas Authors. All rights reserved.
+// This source code is licensed under the Apache 2.0 license found
+// in the LICENSE file in the root directory of this source tree.
+
+package ci
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sort"
+
+	"ariga.io/atlas/sql/migrate"
+	"ariga.io/atlas/sql/schema"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/utils/merkletrie"
+)
+
+type (
+	// A ChangeDetector takes a migration directory and splits it into the "base" files (already merged) and new ones.
+	ChangeDetector interface {
+		// DetectChanges splits the files of a migration directory into the "base" files (already merged) and new ones.
+		DetectChanges(context.Context) ([]migrate.File, []migrate.File, error)
+	}
+	// A ChangeLoader takes a set of migration files and will create multiple schema.Changes out of it.
+	// It will also label migration files as either "generated" or "handcrafted".
+	ChangeLoader interface {
+		// LoadChanges converts each of the given migration files into one Changes.
+		LoadChanges(context.Context, []migrate.File) ([]Changes, error)
```

Comment:
Something I see now, I think the returned result should be "changes per file". This gives better context on what changes are couples and what not.

Reply:
`Changes` is defined as  `type Changes []Change`. Therefire, `[]Changes` are "changes per file". :-)

Reply:
Got it now. Nice.